cpp/include/raft/neighbors/detail/cagra/search_multi_kernel.cuh

```diff
@@ -597,6 +598,95 @@ void set_value_batch(T* const dev_ptr,
     <<<grid_size, block_size, 0, cuda_stream>>>(dev_ptr, ld, val, count, batch_size);
 }
 
+template <class ValT>
+inline void _find_topk(raft::resources const& handle,
+                       uint32_t topK,
+                       uint32_t sizeBatch,
+                       uint32_t numElements,
+                       const float* inputKeys,  // [sizeBatch, ldIK,]
+                       uint32_t ldIK,           // (*) ldIK >= numElements
+                       const ValT* inputVals,   // [sizeBatch, ldIV,]
+                       uint32_t ldIV,           // (*) ldIV >= numElements
+                       float* outputKeys,       // [sizeBatch, ldOK,]
+                       uint32_t ldOK,           // (*) ldOK >= topK
+                       ValT* outputVals,        // [sizeBatch, ldOV,]
+                       uint32_t ldOV,           // (*) ldOV >= topK
+                       void* workspace,
+                       bool sort,
+                       uint32_t* hints)
+{
+  auto stream = resource::get_cuda_stream(handle);
+
+  // _cuann_find_topk right now is limited to a max-k of 1024.
+  // RAFT has a matrix::select_k function - which handles arbitrary sized values of k,
+  // but doesn't accept strided inputs unlike _cuann_find_topk
+  // The multi-kernel search path requires strided access - since its cleverly allocating memory
+  // (layout described in the search_plan_impl function below), such that both the
+  // neighbors and the internal_topk are adjacent - in a double buffered format.
+  // Since this layout doesn't work with the matrix::select_k code - we have to copy
+  // over to a contiguous (non-strided) access to handle topk larger than 1024, and
+  // potentially also copy back to a strided layout afterwards
+  if (topK <= 1024) {
+    return _cuann_find_topk(topK,
+                            sizeBatch,
+                            numElements,
+                            inputKeys,
+                            ldIK,
+                            inputVals,
+                            ldIV,
+                            outputKeys,
+                            ldOK,
+                            outputVals,
+                            ldOV,
+                            workspace,
+                            sort,
+                            hints,
+                            stream);
+  }
+
+  rmm::device_uvector<float> input_keys_storage(0, stream);
+  rmm::device_uvector<float> output_keys_storage(0, stream);
+  rmm::device_uvector<ValT> input_values_storage(0, stream);
+  rmm::device_uvector<ValT> output_values_storage(0, stream);
```

Comment:
To avoid reconstructing these in every function call, I would recommend to make ` _find_topk` a member function of the `search` class, then we can keep these as member variables.

Reply:
thanks for the suggestion! change is in latest commit